src/Microsoft.AspNetCore.Http/Internal/ServerAddressTemplate.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Globalization;
+
+namespace Microsoft.AspNetCore.Http.Internal
+{
+    public class ServerAddressTemplate
+    {
+        public string Host { get; private set; }
+        public string PathBase { get; private set; }
+        public int Port { get; internal set; }
+        public string Scheme { get; private set; }
+
+        public bool IsUnixPipe
+        {
+            get
+            {
+                return Host.StartsWith(Constants.UnixPipeHostPrefix, StringComparison.Ordinal);
+            }
+        }
+
+        public string UnixPipePath
+        {
+            get
+            {
+                Debug.Assert(IsUnixPipe);
+
+                return Host.Substring(Constants.UnixPipeHostPrefix.Length - 1);
+            }
+        }
+
+        public override string ToString()
+        {
+            if (IsUnixPipe)
+            {
+                return Scheme.ToLowerInvariant() + "://" + Host.ToLowerInvariant();
+            }
+            else
+            {
+                return Scheme.ToLowerInvariant() + "://" + Host.ToLowerInvariant() + ":" + Port.ToString(CultureInfo.InvariantCulture) + "/" + PathBase.ToString(CultureInfo.InvariantCulture);
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            return ToString().GetHashCode();
+        }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as ServerAddressTemplate;
+            if (other == null)
+            {
+                return false;
+            }
+            return string.Equals(Scheme, other.Scheme, StringComparison.OrdinalIgnoreCase)
+                && string.Equals(Host, other.Host, StringComparison.OrdinalIgnoreCase)
+                && Port == other.Port;
+        }
+
+        public static ServerAddressTemplate ParseFromUrl(string url)
```

Comment:
To make it clearer that these URLs have a template syntax, how about renaming `ParseFromUrl` to just `Parse` and rename the `url` parameter to `template`?